### PR TITLE
Add `type_string()` utility

### LINKED ISSUE #1860: Add `typeofname()` to get a name of a type rather than its enumerated value
**Describe the project you are working on:**

Any project will do currently tic tac toe game. 

**Describe the problem or limitation you are having in your project:**

if you want to find out what type variable you have you can use typeof() method but it returns enum number. 

    var my_bool: bool = true
    print(typeof(my_bool) # prints 1

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

    I would like to be able to do this instead

    var my_bool: bool = true
    print(typeofname(my_bool) # prints "bool"

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

No idea how it would work with code in c++ 

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Used fairly often and likely there is a work around but in my opinion 99% when user uses typeof() they want to know a name of type not it  enumerated value

**Is there a reason why this should be core and not an add-on in the asset library?:**

Adds function to core functionality.

### PR DESCRIPTION
**Pull request by Kubulambula**
Closes [#1860](https://github.com/godotengine/godot-proposals/issues/1860)

Adds `typeof_string()` method, using `Variant::get_type_name()` effectively just exposing this method to GDScript.

This method goes hand in hand with the already implemented `error_string()` utility. Both of these methods are extremely useful for plugin development where user interaction is expected.

The proposal describes a method that prints the name of the variable directly without the `typeof()` step between. But I believe that this makes this function more versatile and a method like that can be simply made like this:
```gdscript
func typeofname(x):
    return type_string(typeof(x))
```

### CODE REVIEW

**Comment by MewPurPur** [comment]
Regarding the PR itself: I love this feature and I'll be using it a ton in debugging. But since 4.0 is in feature freeze and this doesn't fix any bugs, it will likely not make the cut for now.

File: doc/classes/@GlobalScope.xml
Comment (MewPurPur) [inline doc/classes/@GlobalScope.xml:1213]: You don't need string formatting or looping through an array to explain this concept.

Comment (MewPurPur) [inline doc/classes/@GlobalScope.xml:1213]: You don't need string formatting or looping through an array to explain this concept. Keep it as simple as possible.

Comment (Kubulambula) [reply doc/classes/@GlobalScope.xml:1213]: I specifically did it this way to avoid this function nesting, that might look intimidating at first.
```suggestion
				print(typeof_string(typeof(values[0]))) # Prints int
				print(typeof_string(typeof(values[1]))) # Prints float
				print(typeof_string(typeof(values[1]))) # Prints String
				print(typeof_string(typeof(values[3]))) # Prints bool
```
The solution might be to use more variables, but this would make the code block a lot longer and harder to follow
```suggestion
				var a = 1
				var b = "Hello"
				var c = true
				
				var a_type = typeof(a)
				var b_type = typeof(b)
				var c_type = typeof(c)
				
				print(typeof_string(a_type)) # Prints int
				print(typeof_string(b_type)) # Prints String
				print(typeof_string(c_type)) # Print bool
```

Comment (MewPurPur) [reply doc/classes/@GlobalScope.xml:1213]: Yes, this is better. 2 variables suffice, really.  Also, since the method returns a string, wrap it with quotation marks inside the comments, like `# Prints "String"`

Comment (Kubulambula) [reply doc/classes/@GlobalScope.xml:1213]: What do you think about this?

```suggestion
				func _ready():
					add_one(1) # Prints 2
					add_one(3.0) # Prints 4
					add_one("7") # Prints "Can't add one to String"
					
				func add_one(x):
			                var type = typeof(x)
					if type == TYPE_INT or type == TYPE_FLOAT:
					        print(x + 1)
					else:
					        print("Can't add one to ", typeof_string(type))
				
```

Comment (MewPurPur) [reply doc/classes/@GlobalScope.xml:1213]: Too complicated. I just want to know how `typeof_string()` works, this example forces me to think about irrelevant things, which is massively discouraged in documentation. How about just this?

```
typeof_string(TYPE_INT)  # Prints "int"
typeof_string(TYPE_STRING)  # Prints "String"
```

Comment (Kubulambula) [reply doc/classes/@GlobalScope.xml:1213]: And what about this?

```gdscript
print(TYPE_INT) # Prints 2
print(typeof_string(TYPE_INT)) # Prints "int"
print(typeof_string(TYPE_STRING))  # Prints "String"
```

Comment (MewPurPur) [reply doc/classes/@GlobalScope.xml:1213]: Oop I was referring to the old one! Well, we practically reached the same endpoint then.

Don't forget to amend your commits.

Comment (Kubulambula) [reply doc/classes/@GlobalScope.xml:1213]: Thanks for the feedback ❤️

**Comment by MewPurPur** [comment]
Since this is a small feature, the commits need to be squashed into one. Here's a link to our tutorial on this:

https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by Kubulambula** [comment]
> Since this is a small feature, the commits need to be squashed into one. Here's a link to our tutorial on this: https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

Thanks! I always find a way to mess my branch up

**Comment by 20milliliter** [comment]
@Kubulambula Is this PR ready? I would love this feature, and it could make 4.1.

**Comment by Kubulambula** [comment]
> @Kubulambula Is this PR ready? I would love this feature, and it could make 4.1.

Yes it should be ready as it does not change anything and just exposes an already present function and is just waiting for approval.

**Comment by Kubulambula** [comment]
After looking at it though, maybe `type_string()` would be more fitting?

Because you get the name of type that you get from `typeof()` and `typeof_string()` implies that is gets the string type of something.

Any thoughts?

**Comment by 20milliliter** [comment]
I think it would, too. `type_string()` would make it consistent with the [`error_string()` method](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#class-globalscope-method-error-string).

File: doc/classes/@GlobalScope.xml
Comment (MewPurPur) [inline doc/classes/@GlobalScope.xml:1199]: Don't forget punctuation here. Below too.

**Comment by MewPurPur** [comment]
This is clearly the correct implementation, I think it's a matter of whether we approve of the feature and when we got to this.

File: core/variant/variant_utility.cpp
Comment (bitsawer) [inline core/variant/variant_utility.cpp:754]: I wonder if it would be better to simply return an empty String in this error situation? Would make error checking a bit easier in cases where you might get an invalid type from somewhere.

Comment (Kubulambula) [reply core/variant/variant_utility.cpp:754]: I agree with this, but because of the similarity with the [`error_string()`](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#class-globalscope-method-error-string) function, I wanted to keep consistency as it returns `"(invalid error code)"` string on failure.

Comment (Kubulambula) [reply core/variant/variant_utility.cpp:754]: Also it can be only used by giving it one `int` parameter and the only way of receiving a bad `Enum` value is by the user directly passing in `int` values, instead of receiving them from [`typeof()`](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#class-globalscope-method-typeof), so that is why I think the [`error_string()`](https://docs.godotengine.org/en/stable/classes/class_@globalscope.html#class-globalscope-method-error-string) works like that.

File: core/variant/variant_utility.cpp
Comment (dalexeev) [inline core/variant/variant_utility.cpp:778]:
```suggestion
#ifdef DEBUG_ENABLED
		ERR_FAIL_INDEX_V_MSG((int)type, (int)Variant::VARIANT_MAX, "<invalid type>", "Invalid type argument to type_string(), use the TYPE_* constants.");
#endif
```

**Comment by dalexeev** [comment]
Need to rebase the branch and resolve conflicts.

**Comment by Kubulambula** [comment]
> Need to rebase the branch and resolve conflicts.

Should be good now.

**Comment by AThousandShips** [comment]
You've incorrectly rebased and instead reset all the commits

**Comment by Kubulambula** [comment]
> You've incorrectly rebased and instead reset all the commits

I am trying 😭

**Comment by AThousandShips** [comment]
I'd suggest copying down your changes, doing `git reset --hard master` and then adding your code back in as a single commit

**Comment by Kubulambula** [comment]
@AThousandShips I think that I somehow fixed it. Thank you for your help and patience with me ❤️

File: doc/classes/@GlobalScope.xml
Comment (AThousandShips) [inline doc/classes/@GlobalScope.xml:3202]:
```suggestion
</class>

```

File: core/variant/variant_utility.cpp
Comment (AThousandShips) [inline core/variant/variant_utility.cpp:870]:
```suggestion
#ifdef DEBUG_ENABLED
```

File: core/variant/variant_utility.cpp
Comment (AThousandShips) [inline core/variant/variant_utility.cpp:872]:
```suggestion
#endif
```

File: core/variant/variant_utility.cpp
Comment (AThousandShips) [inline core/variant/variant_utility.cpp:873]:
```suggestion
String VariantUtilityFunctions::type_string(Variant::Type p_type) {
#ifdef DEBUG_ENABLED
	ERR_FAIL_INDEX_V_MSG((int)p_type, (int)Variant::VARIANT_MAX, "<invalid type>", "Invalid type argument to type_string(), use the TYPE_* constants.");
#endif
	return Variant::get_type_name(p_type);
```

File: core/variant/variant_utility.h
Comment (AThousandShips) [inline core/variant/variant_utility.h:131]:
```suggestion
	static String type_string(Variant::Type p_type);
```

**Review by dalexeev — COMMENTED** [review]
Could you please remove the `variant_get_type_name()` function in GDScript tests by the way:

https://github.com/godotengine/godot/blob/7e67b496ff7e35f66b88adcbdd5b252d01739cbb/modules/gdscript/tests/scripts/utils.notest.gd#L69-L76

And replace it here:

https://github.com/godotengine/godot/blob/7e67b496ff7e35f66b88adcbdd5b252d01739cbb/modules/gdscript/tests/scripts/utils.notest.gd#L20

```diff
-	return variant_get_type_name(property.type)
+	return type_string(property.type)
```

File: doc/classes/@GlobalScope.xml
Comment (dalexeev) [inline doc/classes/@GlobalScope.xml:1395]:
```suggestion
				print(type_string(TYPE_INT)) # Prints "int".
				print(type_string(TYPE_STRING)) # Prints "String".
```

Comment (dalexeev) [inline doc/classes/@GlobalScope.xml:1395]:
```suggestion
				print(TYPE_INT) # Prints 2.
				print(type_string(TYPE_INT)) # Prints "int".
				print(type_string(TYPE_STRING)) # Prints "String".
```

**Review by dalexeev — APPROVED** [review]
Looks good to me!

File: core/variant/variant_utility.cpp
Comment (YuriSizov) [inline core/variant/variant_utility.cpp:872]: It's probably worth keeping this in non-debug builds, since this is mostly a developer utility anyway.

Comment (akien-mga) [inline core/variant/variant_utility.cpp:872]: I would keep this in release builds, we don't usually exclude such simple checks from release, unless on a very hot path (which doesn't seem to be the case here).

**Review by YuriSizov — APPROVED** [review]
Other than the note, it looks good!

**Review by akien-mga — APPROVED** [review]
Seems useful to me.

File: core/variant/variant_utility.cpp
Comment (Kubulambula) [reply core/variant/variant_utility.cpp:872]: > I would keep this in release builds, we don't usually exclude such simple checks from release, unless on a very hot path (which doesn't seem to be the case here).

Do you mean just getting rid of the `#ifdef` like this?

```suggestion
ERR_FAIL_INDEX_V_MSG((int)p_type, (int)Variant::VARIANT_MAX, "<invalid type>", "Invalid type argument to type_string(), use the TYPE_* constants.");
```

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**
